psi4/driver/p4util/solvers.py

```diff
@@ -304,3 +308,591 @@ def extrapolate(self, out=None):
+                iter_info['done'] = False
+                Qk = engine.precondition(Rk, lam_k)
+                new_vecs.append(Qk)
+
+        _diag_print_info(print_name, iter_info, verbose)
+        stats.append(iter_info.copy())
+
+        # finished
+        if iter_info['done']:
+            _diag_print_converged(print_name, stats, best_eigvals, verbose)
+            break
+        elif iter_info['collapse']:
+            vecs = best_eigvecs
+        else:
+            vecs = _gs_orth(engine, vecs, new_vecs, schmidt_add_tol)
+
+    # If we get down here  we have exceeded max iterations without convergence, return None + stats, (let caller raise the error)
+    return best_eigvals, best_eigvecs, stats
+
+
+def hamiltonian_solver(engine,
+                       guess,
+                       e_tol=1.0E-6,
+                       r_tol=1.0E-8,
+                       nroot=1,
+                       max_vecs_per_root=20,
+                       maxiter=100,
+                       verbose=1,
+                       schmidt_add_tol=1.0e-8):
+    """
+    Finds the smallest eigenvalues and associated right and left hand eigenvectors of a large real Hamiltonian eigenvalue problem
+    emulated through an engine.
+
+    A hamiltonian EVP has the structure with A, B of some large dimension N the problem is 2Nx2N:
+    [A  B][X]  = [1   0](w)[X]
+    [B  A][Y]    [0  -1](w)[Y]
+
+    Which can be written as the NxN, non-hermitian EVP:
+    (A+B)(A-B)(X+Y) = w^2(X+Y)
+
+
+    With left-hand eigenvectors:
+    (X-Y)(A-B)(A+B) = w^2(X-Y)
+
+    if (A-B) is positive definite, we can transform the problem to arrive at the hermitian NxN EVP:
+    (A-B)^1/2(A+B)(A-B)^1/2 = w^2 T
+
+    Where T = (A-B)^-1/2(X+Y).
+
+    We use a Davidson like iteration where we transform (A+B) (H1) and (A-B) (H2) in to the subspace defined by the trial vectors.
+    The subspace analog of the NxN hermitian EVP is diagonalized and left (X-Y) and right (X+Y) eigenvectors of the NxN
+    non-hermitian EVP are approximated. Residual vectors are formed for both and the guess space is augmented with two
+    correction vectors per iteration.
+
+
+
+    Parameters
+    -----------
+    engine : object
+       The engine drive all operations involving data structures that have one "large" dimension. The methods it is required
+       to provide and their signatures are detailed below.
+    guess : list {engine dependent}
+       At least `nroot` initial expansion vectors
+    e_tol : float
+        Convergence tolerance for eigenvalues
+    r_tol : float
+        Convergence tolerance for residual vectors
+    nroot : int
+        Number of roots desired
+    maxiter : int
+        The maximum number of iterations
+    schmidt_add_tol : float
+        Correction vectors must have norm larger than this value to be added to the guess space
+    verbose : int
+        The amount of logging info to print (0 -> none, 1 -> some, >1 -> everything)
+
+
+    Requirements of engine
```

Comment:
Since this is duplicated, maybe you could just point to a `engine` base class?

Reply:
This would be much cleaner and the Engine requirements are duplicated yet again in the `scf_products` module docstring which could be replaced with a link.